pkg/fqdn/service/service_test.go

```diff
@@ -0,0 +1,244 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/netip"
+	"testing"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/hivetest"
+	"github.com/cilium/hive/job"
+	"github.com/spf13/pflag"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/cilium/cilium/pkg/clustermesh/types"
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/endpoint"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/fqdn/messagehandler"
+	"github.com/cilium/cilium/pkg/fqdn/namemanager"
+	"github.com/cilium/cilium/pkg/hive"
+	health "github.com/cilium/cilium/pkg/hive/health/types"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/policy/api"
+	"github.com/cilium/cilium/pkg/testutils"
+	testidentity "github.com/cilium/cilium/pkg/testutils/identity"
+	"github.com/cilium/cilium/pkg/time"
+
+	pb "github.com/cilium/cilium/api/v1/standalone-dns-proxy"
+)
+
+func TestFQDNDataServer(t *testing.T) {
+	// testutils.PrivilegedTest(t)
+
+	test := map[string]struct {
+		port                     int
+		enableL7Proxy            bool
+		enableStandaloneDNSProxy bool
+		serverPort               int
+		err                      error
+	}{
+		"Successfully running the server": {
+			port:                     1234,
+			enableL7Proxy:            true,
+			enableStandaloneDNSProxy: true,
+			serverPort:               40045,
+			err:                      nil,
+		},
+		"Failure on running the server due to invalid server port": {
+			port:                     1234,
+			serverPort:               0,
+			enableStandaloneDNSProxy: true,
+			enableL7Proxy:            true,
+			err:                      errors.New("listen tcp: address -1: invalid port"),
+		},
+	}
+
+	for scenario, tt := range test {
+		t.Run(scenario, func(t *testing.T) {
+
+			em := endpointmanager.New(&dummyEpSyncher{}, nil, nil, nil, nil)
+			pr := policy.NewPolicyRepository(hivetest.Logger(t), nil, nil, nil, nil, api.NewPolicyMetricsNoop())
+			ipc := ipcache.NewIPCache(&ipcache.Configuration{
```

Comment:
It might make sense to avoid depending on these heavy components directly by defining internal interfaces that only include what is actually needed and then having adapters in `cell.go` that bridge the external type into the internal interface (e.g. `cell.ProvidePrivate(func(ipc *ipcache.IPCache) ipcacheAddListener { return ipc })`) . We can then use the constructor directly here in the test instead of using `Cell` to inject mocks.

If we actually want to go through these components in these tests and not just unit test the `FQDNDataServer` then this is of course fine.

Reply:
I intend to use these for integration testing once we have more code added.